xbmc/pvr/providers/PVRProvider.h

```diff
@@ -0,0 +1,217 @@
+
+class CPVRProvider final : public ISerializable
+{
+public:
+  CPVRProvider(int iUniqueId, int iClientId);
+  CPVRProvider(const PVR_PROVIDER& provider, int iClientId);
+  CPVRProvider(int iClientId,
+               const std::string& addonProviderName,
+               const std::string& addonIconPath,
+               const std::string& addonThumbPath);
+
+  bool operator==(const CPVRProvider& right) const;
+  bool operator!=(const CPVRProvider& right) const;
+
+  void Serialize(CVariant& value) const override;
+
+  /*!
+   * @brief The database id of this provider
+   *
+   * A unique identifier for this provider.
+   * It can be used to find the same provider on this clients backend
+   *
+   * @return The database id of this provider
+   */
+  int GetProviderId() const;
+
+  /*!
+   * @brief Set the database id of this provider
+   * @param iProviderId The new ID.
+   * @return True if the something changed, false otherwise.
+   */
+  bool SetProviderId(int iProviderId);
+
+  /*!
+   * @brief A unique identifier for this provider.
+   *
+   * A unique identifier for this provider.
+   * It can be used to find the same provider on this clients backend
+   *
+   * @return The Unique ID.
+   */
+  int GetUniqueId() const;
+
+  /*!
+   * @return The identifier of the client that supplies this provider.
+   */
+  int GetClientId() const;
+
+  /*!
+   * @return The name of the provider. Can be user provided or the backend name
+   */
+  std::string GetProviderName() const;
+
+  /*!
+   * @brief Set the name of the provider.
+   * @param providerName The new name of the provider.
+   * @return True if the something changed, false otherwise.
+   */
+  bool SetProviderName(const std::string& iProviderName);
+
+  /*!
+   * @brief Checks whether this provider has a type other than
+   *        PVR_PROVIDER_TYPE_UNKNOWN
+   * @return True if this provider has a type other than unknown, false otherwise
+   */
+  bool HasProviderType() const { return m_providerType != PVR_PROVIDER_TYPE_UNKNOWN; }
+
+  /*!
+   * @brief Gets the type of this provider.
+   * @return the type of this provider.
+   */
+  PVR_PROVIDER_TYPE GetProviderType() const;
+
+  /*!
+   * @brief Sets the type of this provider.
+   * @param providerType the new provider type.
+   * @return True if the something changed, false otherwise.
+   */
+  bool SetProviderType(PVR_PROVIDER_TYPE providerType);
+
+  /*!
+   * @brief Get the iconpath for this provider's icon
+   * @return iconpath for this provider's icon
+   */
+  std::string GetIconPath() const;
+
+  /*!
+   * @brief Set the path for this icon
+   * @param strIconPath The new path of the icon.
+   * @return true if the icon path was updated successfully
+   */
+  bool SetIconPath(const std::string& strIconPath);
+
+  /*!
+   * @brief Get this provider's country codes (ISO 3166). Separated by
+   *        PROVIDER_STRING_TOKEN_SEPARATOR.
+   * @return This provider's country codes.
+   */
+  std::string GetCountry() const;
```

Comment:
would it make sense to de-tokenize internally and only to have a method `std::vector<std::string> GetCountry()`, for convenience reasons?

And why is it called GetCountry (singular) if there can be more than one country code?

Reply:
We have functions called GetWriter() and GetDirector(), I can change it easy enough though.

Reply:
We can I guess, but we would need to expose both split and unsplit versions. For instance when we want to read/write to the DB we want the plain string and it would mean tokenising and detokensing externally too. Although, not necessarily something we need right now.

Reply:
> We have functions called GetWriter() and GetDirector(), I can change it easy enough though.

We should fix that in another PR for writer and director, but not copy that "bug" for new code.

Reply:
> We can I guess, but we would need to expose both split and unsplit versions. 

The way to go, imo. Keep internals internal. Do not expose implementation details.

Reply:
> > We can I guess, but we would need to expose both split and unsplit versions.
> 
> The way to go, imo. Keep internals internal. Do not expose implementation details.

I agree but not entirely seeing how this is possible in this case due to the DB requiring a single value. It guess if we use vectors for the standard getters/setters and then `GetCountriesDBString(`) and `SetCountriesFromDBString()`? Or would it be better to make the tokensie/detokenise methods public so the DB can use them? I'm leaning toward the latter.

Reply:
Interesting. If prefer the former.

Reply:
Ok, the former it is.

Reply:
Done